README.md
<!-- omit in toc -->
# go-nagios

Shared Golang package for Nagios plugins

[![Latest Release](https://img.shields.io/github/release/atc0005/go-nagios.svg?style=flat-square)](https://github.com/atc0005/go-nagios/releases/latest)
[![Go Reference](https://pkg.go.dev/badge/github.com/atc0005/go-nagios.svg)](https://pkg.go.dev/github.com/atc0005/go-nagios)
[![go.mod Go version](https://img.shields.io/github/go-mod/go-version/atc0005/go-nagios)](https://github.com/atc0005/go-nagios)
[![Lint and Build](https://github.com/atc0005/go-nagios/actions/workflows/lint-and-build.yml/badge.svg)](https://github.com/atc0005/go-nagios/actions/workflows/lint-and-build.yml)
[![Project Analysis](https://github.com/atc0005/go-nagios/actions/workflows/project-analysis.yml/badge.svg)](https://github.com/atc0005/go-nagios/actions/workflows/project-analysis.yml)
[![Push Validation](https://github.com/atc0005/go-nagios/actions/workflows/push-validation.yml/badge.svg)](https://github.com/atc0005/go-nagios/actions/workflows/push-validation.yml)

<!-- omit in toc -->
## Table of contents

- [Status](#status)
- [Overview](#overview)
- [Features](#features)
- [Changelog](#changelog)
- [Examples](#examples)
  - [Import this library](#import-this-library)
  - [Use only the provided constants](#use-only-the-provided-constants)
  - [Basic plugin structure](#basic-plugin-structure)
  - [Use a branding callback](#use-a-branding-callback)
  - [Override section header/label text](#override-section-headerlabel-text)
  - [Omit Errors, Thresholds sections](#omit-errors-thresholds-sections)
  - [Collect and emit Performance Data](#collect-and-emit-performance-data)
- [License](#license)
- [References](#references)

## Status

Alpha quality.

This codebase is subject to change without notice and may break client code
that depends on it. You are encouraged to [vendor](#references) this package
if you find it useful until such time that the API is considered stable.

## Overview

This package contains common types and package-level variables used when
developing Nagios plugins. The intent is to reduce code duplication between
various plugins and help reduce typos associated with literal strings.

## Features

- Nagios state constants
  - state labels (e.g., `StateOKLabel`)
  - state exit codes (e.g., `StateOKExitCode`)
- Nagios `CheckOutputEOL` constant
  - provides a consistent newline format for both Nagios Core and Nagios XI
    (and presumably other similar monitoring systems)
- Nagios `ServiceState` type
  - simple label and exit code "wrapper"
  - useful in client code as a way to map internal check results to a Nagios
    service state value
- Supports "branding" callback function to display application name,
  version, or other information as a "trailer" for check results provided to
  Nagios
  - this could be useful for identifying what version of a plugin determined
    the service or host state to be an issue
- Panics from client code are captured and reported
  - panics are surfaced as `CRITICAL` state
  - service output and error details are overridden to panic prominent
- Optional support for emitting performance data generated by plugins
  - NOTE: The implementation of this support is not yet stable and may
    change in the future. See GH-81 and GH-92 for additional details.
- Support for collecting multiple errors from client code
- Support for explicitly omitting Errors section in `LongServiceOutput`
  - this section is automatically omitted if no errors were recorded (by
    client code or panic handling code)
- Support for explicitly omitting Thresholds section in `LongServiceOutput`
  - this section is automatically omitted if no thresholds were specified by
    client code
- Automatically omit `LongServiceOutput` section if not specify by client code
- Support for overriding text used for section headers/labels

## Changelog

See the [`CHANGELOG.md`](CHANGELOG.md) file for the changes associated with
each release of this application. Changes that have been merged to `master`,
but not yet an official release may also be noted in the file under the
`Unreleased` section. A helpful link to the Git commit history since the last
official release is also provided for further review.

## Examples

### Import this library

Add this line to your imports like so:

```golang
package main

import (
  "fmt"
  "log"
  "os"

  "github.com/atc0005/go-nagios"
)
```

and pull in a specific version of this library that you'd like to use.

```console
go get github.com/atc0005/go-nagios@v0.9.0
```

Alternatively, you can use the latest stable tag available to get started:

```console
go get github.com/atc0005/go-nagios@latest
```

### Use only the provided constants

After you've imported this library, reference the exported data types as you
would from any other package. In this example, we reference a specific exit
code for the `OK` state:

```golang
fmt.Println("OK: All checks have passed")
os.Exit(nagios.StateOKExitCode)
```

You can also use the provided state "labels" to avoid using literal string
state values (recommended):

```golang
fmt.Printf(
    "%s: All checks have passed%s",
    nagios.StateOKLabel,
    nagios.CheckOutputEOL,
)

os.Exit(nagios.StateOKExitCode)
```

### Basic plugin structure

First, create an instance of the `ExitState` type and immediately `defer`
`ReturnCheckResults()` so that it runs as the last step in your client code.

Also, avoid calling `os.Exit()` directly from your code. If you do, this
library is unable to function properly; this library expects that *it* will
handle calling `os.Exit()` with the required exit code (and specifically
formatted output).

Also, if you do not `defer` `ReturnCheckResults()` immediately any other
deferred functions in your client code *will not run*.

Here we're optimistic and we are going to note that all went well.

```golang
package main

import (
  // ...
)

func main() {

    var nagiosExitState = nagios.ExitState{
        LastError:         nil,
        ExitStatusCode:    nagios.StateOKExitCode,
    }

    defer nagiosExitState.ReturnCheckResults()

    // more stuff here

    nagiosExitState.ServiceOutput = certs.OneLineCheckSummary(
        nagios.StateOKLabel,
        certChain,
        certsSummary.Summary,
    )

    nagiosExitState.LongServiceOutput := certs.GenerateCertsReport(
        certChain,
        certsExpireAgeCritical,
        certsExpireAgeWarning,
    )

```

For handling error cases, the approach is roughly the same, only you call
`return` explicitly to end execution of the client code and allow deferred
functions to run.

### Use a branding callback

In this example, we'll make a further assumption that you have a `config`
value with an `EmitBranding` field to indicate whether the user/sysadmin has
opted to emit branding information.

```golang
package main

import (
  // ...
)

func main() {

    var nagiosExitState = nagios.ExitState{
        LastError:         nil,
        ExitStatusCode:    nagios.StateOKExitCode,
    }

    defer nagiosExitState.ReturnCheckResults()

    // ...

    if config.EmitBranding {
      // If enabled, show application details at end of notification
      nagiosExitState.BrandingCallback = Branding("Notification generated by ")
    }

    // ...

}
```

the `Branding` function might look something like this:

```golang
// Branding accepts a message and returns a function that concatenates that
// message with version information. This function is intended to be called as
// a final step before application exit after any other output has already
// been emitted.
func Branding(msg string) func() string {
    return func() string {
        return strings.Join([]string{msg, Version()}, "")
    }
}
```

but you could just as easily create an anonymous function as the callback:

```golang
package main

import (
  // ...
)

func main() {

    var nagiosExitState = nagios.ExitState{
        LastError:         nil,
        ExitStatusCode:    nagios.StateOKExitCode,
    }

    defer nagiosExitState.ReturnCheckResults()

    if config.EmitBranding {
        // If enabled, show application details at end of notification
        nagiosExitState.BrandingCallback = func(msg string) func() string {
            return func() string {
                return "Notification generated by " + msg
            }
        }("HelloWorld")
    }

    // ...

}
```

### Override section header/label text

In this example, we override the default text with values that better fit our
use case.

```golang
package main

import (
  // ...
)

func main() {

    var nagiosExitState = nagios.ExitState{
        LastError:         nil,
        ExitStatusCode:    nagios.StateOKExitCode,
    }

    defer nagiosExitState.ReturnCheckResults()

    // Override default section headers with our custom values.
    nagiosExitState.SetErrorsLabel("VALIDATION ERRORS")
    nagiosExitState.SetDetailedInfoLabel("VALIDATION CHECKS REPORT")

    // more stuff here

    nagiosExitState.ServiceOutput = certs.OneLineCheckSummary(
        nagios.StateOKLabel,
        certChain,
        certsSummary.Summary,
    )

    nagiosExitState.LongServiceOutput := certs.GenerateCertsReport(
        certChain,
        certsExpireAgeCritical,
        certsExpireAgeWarning,
    )

```

### Omit Errors, Thresholds sections

In this example, we hide or omit the Errors and Thresholds sections entirely.

```golang
package main

import (
  // ...
)

func main() {

    var nagiosExitState = nagios.ExitState{
        LastError:         nil,
        ExitStatusCode:    nagios.StateOKExitCode,
    }

    defer nagiosExitState.ReturnCheckResults()

    // Hide/Omit these sections from plugin output
    nagiosExitState.HideErrorsSection()
    nagiosExitState.HideThresholdsSection()

    // more stuff here

    nagiosExitState.ServiceOutput = certs.OneLineCheckSummary(
        nagios.StateOKLabel,
        certChain,
        certsSummary.Summary,
    )

    nagiosExitState.LongServiceOutput := certs.GenerateCertsReport(
        certChain,
        certsExpireAgeCritical,
        certsExpireAgeWarning,
    )

```

### Collect and emit Performance Data

This example provides plugin runtime via a deferred anonymous function:

```golang
package main

import (
  // ...
)

func main() {

    // Start the timer. We'll use this to emit the plugin runtime as a
    // performance data metric.
    pluginStart := time.Now()

    var nagiosExitState = nagios.ExitState{
        LastError:         nil,
        ExitStatusCode:    nagios.StateOKExitCode,
    }

    // defer this from the start so it is the last deferred function to run
    defer nagiosExitState.ReturnCheckResults()

    // Collect last minute details just before ending plugin execution.
    defer func(exitState *nagios.ExitState, start time.Time, logger zerolog.Logger) {

        // Record plugin runtime, emit this metric regardless of exit
        // point/cause.
        runtimeMetric := nagios.PerformanceData{
            Label: "time",
            Value: fmt.Sprintf("%dms", time.Since(start).Milliseconds()),
        }
        if err := exitState.AddPerfData(false, runtimeMetric); err != nil {
            zlog.Error().
                Err(err).
                Msg("failed to add time (runtime) performance data metric")
        }

        // Annotate errors (if applicable) with additional context to aid in
        // troubleshooting.
        nagiosExitState.Errors = annotateError(logger, nagiosExitState.Errors...)
    }(&nagiosExitState, pluginStart, cfg.Log)

    // more stuff here

    // This example also assumes that the check results indicate success. You
    // may opt to use a "IsOK()" style check and a switch statement to
    // have the below execute within a default branch.
    nagiosExitState.ServiceOutput = certs.OneLineCheckSummary(
        nagios.StateOKLabel,
        certChain,
        certsSummary.Summary,
    )

    nagiosExitState.LongServiceOutput := certs.GenerateCertsReport(
        certChain,
        certsExpireAgeCritical,
        certsExpireAgeWarning,
    )

    // ...

}
```

and this example provides multiple performance data values explicitly:

```golang
package main

import (
  // ...
)

func main() {

    // Start the timer. We'll use this to emit the plugin runtime as a
    // performance data metric.
    pluginStart := time.Now()

    var nagiosExitState = nagios.ExitState{
        LastError:         nil,
        ExitStatusCode:    nagios.StateOKExitCode,
    }

    // defer this from the start so it is the last deferred function to run
    defer nagiosExitState.ReturnCheckResults()

    // more stuff here

    pd := []nagios.PerformanceData{
        {
            Label: "time",
            Value: fmt.Sprintf("%dms", time.Since(start).Milliseconds()),
        },
        {
            Label: "datacenters",
            Value: fmt.Sprintf("%d", len(dcs)),
        },
        {
            Label: "triggered_alarms",
            Value: fmt.Sprintf("%d", len(triggeredAlarms)),
        },
    }

    if err := nagiosExitState.AddPerfData(false, pd...); err != nil {
        log.Error().
            Err(err).
            Msg("failed to add performance data")
    }

    // This example also assumes that the check results indicate success. You
    // may opt to use a "IsOK()" style check and a switch statement to
    // have the below execute within a default branch.
    nagiosExitState.ServiceOutput = certs.OneLineCheckSummary(
        nagios.StateOKLabel,
        certChain,
        certsSummary.Summary,
    )

    nagiosExitState.LongServiceOutput := certs.GenerateCertsReport(
        certChain,
        certsExpireAgeCritical,
        certsExpireAgeWarning,
    )

    // ...

}
```

This example drops the deferred handling of the plugin runtime metric in order
to illustrate how it would look if handled directly alongside other metrics.

See these issues for further details:

- [GH-81](https://github.com/atc0005/go-nagios/issues/81)
- [GH-92](https://github.com/atc0005/go-nagios/issues/92)

## License

From the [LICENSE](LICENSE) file:

```license
MIT License

Copyright (c) 2020 Adam Chalkley

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.


CHANGELOG.md
# Changelog

## Overview

All notable changes to this project will be documented in this file.

The format is based on [Keep a
Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to
[Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Please [open an issue](https://github.com/atc0005/go-nagios/issues) for any
deviations that you spot; I'm still learning!.

## Types of changes

The following types of changes will be recorded in this file:

- `Added` for new features.
- `Changed` for changes in existing functionality.
- `Deprecated` for soon-to-be removed features.
- `Removed` for now removed features.
- `Fixed` for any bug fixes.
- `Security` in case of vulnerabilities.

## [Unreleased]

- placeholder

## [v0.12.1] - 2022-12-15

### Fixed

- (GH-172) Replace further ExitState references

## [v0.12.0] - 2022-12-15

### Changed

- (GH-169) Rename primary `ExitState` type to `Plugin`, `New` constructor to
  `NewPlugin`

## [v0.11.0] - 2022-12-14

### Added

- (GH-102) Add constructor to initialize plugin start time, provide `time`
  Performance Data metric "automatically"

## [v0.10.2] - 2022-10-19

### Fixed

- (GH-164) Sort performance data metrics for consistent output

## [v0.10.1] - 2022-10-09

### Changed

- Dependencies
  - `google/go-cmp`
    - `v0.5.8` to `v0.5.9`
- (GH-103) Emit Performance Data on same line as `ServiceOutput` if
  `LongServiceOutput` is empty
- (GH-146) Move examples from README to Example tests
- (GH-154) Refactor GitHub Actions workflows to import logic
- (GH-156) Update README to include go.mod badge
- (GH-157) Prevent duplication of Performance Data metrics

## [v0.10.0] - 2022-09-18

### Added

- (GH-144) Add test to prevent further Service Output interpolation
  regressions
- (GH-145) Add support for setting preferred output target
- (GH-147) Add option to skip os.Exit call at end of plugin execution
- (GH-149) Add initial plugin output test

### Changed

- Dependencies
  - `github/codeql-action`
    - `v2.1.22` to `v2.1.23`

## [v0.9.2] - 2022-09-15

### Changed

- (GH-137) Update project to Go 1.19
- (GH-138) Update Makefile and GitHub Actions Workflows

### Fixed

- (GH-139) `ExitState.ServiceOutput` is (incorrectly) interpreted as
  containing format specifier

## [v0.9.1] - 2022-06-20

### Fixed

- (GH-132) `LongServiceOutput` header is emitted when
  `ExitState.LongServiceOutput` field is empty
- (GH-134) Update lintinstall Makefile recipe
- (GH-135) Fix Go 1.19beta1 gofmt linter warning

## [v0.9.0] - 2022-06-16

### Added

- (GH-107) Add support for overriding section headers/labels
- (GH-117) Add options to explicitly "hide" or omit Thresholds and Errors
  section, automatically hide Detailed Info section if those sections are
  unused
- (GH-120) Allow collecting multiple errors

### Changed

- Dependencies
  - `actions/checkout`
    - `v2.4.0` to `v3`
  - `actions/setup-node`
    - `v2.5.1` to `v3`

- (GH-118) Automatically omit Thresholds and Errors sections if unused
- (GH-112) Expand linting GitHub Actions Workflow to include `oldstable`,
  `unstable` container images
- (GH-113) Switch Docker image source from Docker Hub to GitHub Container
  Registry (GHCR)
- (GH-130) Refresh README, go.doc files for v0.9.0 release

### Fixed

- (GH-121) `err113` linting issues: `do not define dynamic errors, use wrapped
  static errors instead`
- (GH-127) cognitive complexity 32 of func `(*ExitState).ReturnCheckResults`
  is high (> 10) (gocognit)

## [v0.8.2] - 2022-01-01

### Changed

- Dependencies
  - `actions/checkout`
    - `v2.3.5` to `v2.4.0`
  - `actions/setup-node`
    - `v2.4.1` to `v2.5.1`

### Fixed

- (GH-104) Swap nil PerformanceData check for length check
- (GH-109) Nagios XI interprets current `nagios.CheckOutputEOL` value as two
  newlines

## [v0.8.1] - 2021-11-02

### Changed

- Dependencies
  - `actions/checkout`
    - `v2.3.4` to `v2.3.5`

### Fixed

- (GH-98) Performance Data emitted when `ServiceOutput` and
  `LongServiceOutput` empty
- (GH-96) CHANGELOG | Fix typo and clarify perfdata support

## [v0.8.0] - 2021-09-30

### Added

- Add initial support for collecting, formatting & emitting performance data

### Changed

- Dependencies
  - `actions/setup-node`
    - `v2.4.0` to `v2.4.1`

### Fixed

- Fix typo in CHANGELOG release tags list
- Remove duplicate package doc comments

## [v0.7.0] - 2021-09-07

### Added

- (GH-80) Add `ServiceState` type

### Changed

- Dependencies
  - `actions/setup-node`
    - (`v2.2.0` to `v2.4.0`

### Fixed

- (GH-76) Fix typo in doc comment for `ExitStatusCode` field
- (GH-82) Typo in `UNKNOWN` state label text

## [v0.6.1] - 2021-07-15

### Changed

- Panic error message
  - (GH-69) stack trace wrapped with Markdown code fences instead of `pre`
    tags
  - (GH-68) provide complete stack trace instead of snippet
- Documentation
  - Replace GoDoc badge with pkg.go.dev badge
- Dependencies
  - `actions/setup-node`
    - (`v2.1.4` to `v2.2.0`
    - update `node-version` value to always use latest LTS version instead of
      hard-coded version

## [v0.6.0] - 2021-01-14

### Added

- Expose panic error message
  - emit as first line item in pre block above stack trace
  - emit as part of error entry

## [v0.5.3] - 2021-01-05

### Changed

- dependencies
  - `actions/setup-node`
    - `v2.1.2` to `v2.1.4`

### Fixed

- Preformatted `ServiceOutput` string subjected to another (failed) formatting
  operation
- Breadcrumb URL formatting

## [v0.5.2] - 2020-11-08

### Changed

- dependencies
  - `actions/setup-node`
    - `v2.1.1` to `v2.1.2`
  - `actions/checkout`
    - `v2.3.2` to `v2.3.4`

### Fixed

- (*nagios.ExitState).ReturnCheckResults() unintentionally masks or "swallows"
  panics

## [v0.5.1] - 2020-09-22

### Changed

- `ExitState` receiver type is now a pointer for `ReturnCheckResults()` method

### Fixed

- Documentation
  - Update doc comments and README examples to (hopefully) better explain
    method usage

## [v0.5.0] - 2020-09-20

### Changed


nagios.go
// Copyright 2020 Adam Chalkley
//
// https://github.com/atc0005/go-nagios
//
// Licensed under the MIT License. See LICENSE file in the project root for
// full license information.

package nagios

import (
	"errors"
	"fmt"
	"io"
	"os"
	"runtime/debug"
	"strings"
)

// Nagios plugin/service check states. These constants replicate the values
// from utils.sh which is normally found at one of these two locations,
// depending on which Linux distribution you're using:
//
// - /usr/lib/nagios/plugins/utils.sh
// - /usr/local/nagios/libexec/utils.sh
//
// See also http://nagios-plugins.org/doc/guidelines.html
const (
	StateOKExitCode        int = 0
	StateWARNINGExitCode   int = 1
	StateCRITICALExitCode  int = 2
	StateUNKNOWNExitCode   int = 3
	StateDEPENDENTExitCode int = 4
)

// Nagios plugin/service check state "labels". These constants are provided as
// an alternative to using literal state strings throughout client application
// code.
const (
	StateOKLabel        string = "OK"
	StateWARNINGLabel   string = "WARNING"
	StateCRITICALLabel  string = "CRITICAL"
	StateUNKNOWNLabel   string = "UNKNOWN"
	StateDEPENDENTLabel string = "DEPENDENT"
)

// CheckOutputEOL is the newline character(s) used with formatted service and
// host check output. Based on previous testing, Nagios treats LF newlines
// (without a leading space) within the `$LONGSERVICEOUTPUT$` macro as literal
// values instead of parsing them for display purposes.
//
// Using DOS EOL values with fmt.Fprintf() (or fmt.Fprintln()) gives expected
// formatting results in the Nagios Core web UI, but results in double
// newlines in Nagios XI output (see GH-109). Using a UNIX EOL with a single
// leading space appears to give the intended results for both Nagios Core and
// Nagios XI.
const CheckOutputEOL string = " \n"

// Default header text for various sections of the output if not overridden.
const (
	defaultThresholdsLabel   string = "THRESHOLDS"
	defaultErrorsLabel       string = "ERRORS"
	defaultDetailedInfoLabel string = "DETAILED INFO"
)

// Sentinel error collection. Exported for potential use by client code to
// detect & handle specific error scenarios.
var (
	// ErrPanicDetected indicates that client code has an unhandled panic and
	// that this library detected it before it could cause the plugin to
	// abort. This error is included in the LongServiceOutput emitted by the
	// plugin.
	ErrPanicDetected = errors.New("plugin crash/panic detected")

	// ErrPerformanceDataMissingLabel indicates that client code did not
	// provide a PerformanceData value in the expected format; the label for
	// the label/value pair is missing.
	ErrPerformanceDataMissingLabel = errors.New("provided performance data missing required label")

	// ErrPerformanceDataMissingValue indicates that client code did not
	// provide a PerformanceData value in the expected format; the value for
	// the label/value pair is missing.
	ErrPerformanceDataMissingValue = errors.New("provided performance data missing required value")

	// ErrNoPerformanceDataProvided indicates that client code did not provide
	// the expected PerformanceData value(s).
	ErrNoPerformanceDataProvided = errors.New("no performance data provided")
)

// ServiceState represents the status label and exit code for a service check.
type ServiceState struct {

	// Label maps directly to one of the supported Nagios state labels.
	Label string

	// ExitCode is the exit or exit status code associated with a Nagios
	// service check.
	ExitCode int
}

// PerformanceData represents the performance data generated by a Nagios
// plugin.
//
// Plugin performance data is external data specific to the plugin used to
// perform the host or service check. Plugin-specific data can include things
// like percent packet loss, free disk space, processor load, number of
// current users, etc. - basically any type of metric that the plugin is
// measuring when it executes.
type PerformanceData struct {

	// Label is the single quoted text string used as a label for a specific
	// performance data point. The label length is arbitrary, but ideally the
	// first 19 characters are unique due to a limitation in RRD. There is
	// also a limitation in the amount of data that NRPE returns to Nagios.
	//
	// The popular convention used by plugin authors (and official
	// documentation) is to use underscores for separating multiple words. For
	// example, 'percent_packet_loss' instead of 'percent packet loss',
	// 'percentPacketLoss' or 'percent-packet-loss.
	Label string

	// Value is the data point associated with the performance data label.
	//
	// Value is in class [-0-9.] and must be the same UOM as Min and Max UOM.
	// Value may be a literal "U" instead, this would indicate that the actual
	// value couldn't be determined.
	Value string

	// UnitOfMeasurement is an optional unit of measurement (UOM). If
	// provided, consists of a string of zero or more characters. Numbers,
	// semicolons or quotes are not permitted.
	//
	// Examples:
	//
	// 1) no unit specified - assume a number (int or float) of things (eg,
	// users, processes, load averages)
	// 2) s - seconds (also us, ms)
	// 3) % - percentage
	// 4) B - bytes (also KB, MB, TB)
	// 5) c - a continuous counter (such as bytes transmitted on an interface)
	UnitOfMeasurement string

	// Warn is in the range format (see the Section called Threshold and
	// Ranges). Must be the same UOM as Crit. An empty string is permitted.
	//
	// https://nagios-plugins.org/doc/guidelines.html#THRESHOLDFORMAT
	Warn string

	// Crit is in the range format (see the Section called Threshold and
	// Ranges). Must be the same UOM as Warn. An empty string is permitted.
	//
	// https://nagios-plugins.org/doc/guidelines.html#THRESHOLDFORMAT
	Crit string

	// Min is in class [-0-9.] and must be the same UOM as Value and Max. Min
	// is not required if UOM=%. An empty string is permitted.
	Min string

	// Max is in class [-0-9.] and must be the same UOM as Value and Min. Max
	// is not required if UOM=%. An empty string is permitted.
	Max string
}

// Validate performs basic validation of PerformanceData. An error is returned
// for any validation failures.
func (pd PerformanceData) Validate() error {

	// Validate fields
	switch {
	case pd.Label == "":
		return ErrPerformanceDataMissingLabel
	case pd.Value == "":
		return ErrPerformanceDataMissingValue

	// TODO: Expand validation
	// https://nagios-plugins.org/doc/guidelines.html
	default:
		return nil

	}
}

// ExitCallBackFunc represents a function that is called as a final step
// before application termination so that branding information can be emitted
// for inclusion in the notification. This helps identify which specific
// application (and its version) that is responsible for the notification.
type ExitCallBackFunc func() string

// ExitState represents the last known execution state of the
// application, including the most recent error and the final intended plugin
// state.
type ExitState struct {
	// outputSink is the user-specified or fallback target for Nagios plugin
	// output.
	outputSink io.Writer

	// LastError is the last error encountered which should be reported as
	// part of ending the service check (e.g., "Failed to connect to XYZ to
	// check contents of Inbox").
	//
	// Deprecated: Use Errors field or AddError method instead.
	LastError error

	// Errors is a collection of one or more recorded errors to be displayed
	// in LongServiceOutput as a list when ending the service check.
	Errors []error

	// ExitStatusCode is the exit or exit status code provided to the Nagios
	// instance that calls this service check. These status codes indicate to
	// Nagios "state" the service is considered to be in. The most common
	// states are OK (0), WARNING (1) and CRITICAL (2).
	ExitStatusCode int

	// ServiceOutput is the first line of text output from the last service
	// check (i.e. "Ping OK").
	ServiceOutput string

	// LongServiceOutput is the full text output (aside from the first line)
	// from the last service check.
	LongServiceOutput string

	// perfData is the collection of zero or more PerformanceData values
	// generated by a Nagios plugin.
	perfData []PerformanceData

	// WarningThreshold is the value used to determine when the service check
	// has crossed between an existing state into a WARNING state. This value
	// is used for display purposes.
	WarningThreshold string

	// CriticalThreshold is the value used to determine when the service check
	// has crossed between an existing state into a CRITICAL state. This value
	// is used for display purposes.
	CriticalThreshold string

	// thresholdLabel is an optional custom label used in place of the
	// standard text prior to a list of threshold values.
	thresholdsLabel string

	// errorsLabel is an optional custom label used in place of the standard
	// text prior to a list of recorded error values.
	errorsLabel string

	// detailedInfoLabel is an optional custom label used in place of the
	// standard text prior to emitting LongServiceOutput.
	detailedInfoLabel string

	// hideThresholdsSection indicates whether client code has opted to hide
	// the thresholds section, regardless of whether client code previously
	// specified values for display.
	hideThresholdsSection bool

	// hideErrorsSection indicates whether client code has opted to hide the
	// errors section, regardless of whether client code previously specified
	// values for display.
	hideErrorsSection bool

	// BrandingCallback is a function that is called before application
	// termination to emit branding details at the end of the notification.
	// See also ExitCallBackFunc.
	BrandingCallback ExitCallBackFunc
}

// ReturnCheckResults is intended to provide a reliable way to return a
// desired exit code from applications used as Nagios plugins. In most cases,
// this method should be registered as the first deferred function in client
// code. See remarks regarding "masking" or "swallowing" application panics.
//
// Since Nagios relies on plugin exit codes to determine success/failure of
// checks, the approach that is most often used with other languages is to use
// something like Using os.Exit() directly and force an early exit of the
// application with an explicit exit code. Using os.Exit() directly in Go does
// not run deferred functions. Go-based plugins that do not rely on deferring
// function calls may be able to use os.Exit(), but introducing new
// dependencies later could introduce problems if those dependencies rely on
// deferring functions.
//
// Before calling this method, client code should first set appropriate field
// values on the receiver. When called, this method will process them and exit
// with the desired exit code and status output.
//
// To repeat, if scheduled via defer, this method should be registered first;
// because this method calls os.Exit to set the intended plugin exit state, no
// other deferred functions will have an opportunity to run, so register this
// method first so that when deferred, it will be run last (FILO).
//
// Because this method is (or should be) deferred first within client code, it
// will run after all other deferred functions. It will also run before a
// panic in client code forces the application to exit. As already noted, this
// method calls os.Exit to set the plugin exit state. Because os.Exit forces
// the application to terminate immediately without running other deferred
// functions or processing panics, this "masks", "swallows" or "blocks" panics
// from client code from surfacing. This method checks for unhandled panics
// and if found, overrides exit state details from client code and surfaces
// details from the panic instead as a CRITICAL state.
func (es *ExitState) ReturnCheckResults() {

	var output strings.Builder

	// ##################################################################
	// Note: fmt.Println() (and fmt.Fprintln()) has the same issue as `\n`:
	// Nagios seems to interpret them literally instead of emitting an actual
	// newline. We work around that by using fmt.Fprintf() and fmt.Fprint()
	// for output that is intended for display within the Nagios web UI.
	// ##################################################################

	// Check for unhandled panic in client code. If present, override
	// ExitState and make clear that the client code/plugin crashed.
	if err := recover(); err != nil {

		es.AddError(fmt.Errorf("%w: %s", ErrPanicDetected, err))

		es.ServiceOutput = fmt.Sprintf(
			"%s: plugin crash detected. See details via web UI or run plugin manually via CLI.",
			StateCRITICALLabel,
		)

		// Gather stack trace associated with panic.
		stackTrace := debug.Stack()

		// Wrap stack trace details in an attempt to prevent these details
		// from being interpreted as formatting characters when passed through
		// web UI, text, email, Teams, etc. We use Markdown fenced code blocks
		// instead of `<pre>` start/end tags because Nagios strips out angle
		// brackets (due to default `illegal_macro_output_chars` settings).
		es.LongServiceOutput = fmt.Sprintf(
			"```%s%s%s%s%s%s```",
			CheckOutputEOL,
			err,
			CheckOutputEOL,
			CheckOutputEOL,
			stackTrace,
			CheckOutputEOL,
		)

		es.ExitStatusCode = StateCRITICALExitCode

	}

	es.handleServiceOutputSection(&output)

	es.handleErrorsSection(&output)

	es.handleThresholdsSection(&output)

	es.handleLongServiceOutput(&output)

	// If set, call user-provided branding function before emitting
	// performance data and exiting application.
	if es.BrandingCallback != nil {
		fmt.Fprintf(&output, "%s%s%s", CheckOutputEOL, es.BrandingCallback(), CheckOutputEOL)
	}

	es.handlePerformanceData(&output)

	// Emit all collected plugin output using user-specified or fallback
	// output target.
	es.emitOutput(output.String())

	os.Exit(es.ExitStatusCode)
}

// AddPerfData appends provided performance data. Validation is skipped if
// requested, otherwise an error is returned if validation fails. Validation
// failure results in no performance data being appended.
//
// Client code may wish to disable validation if performing this step
// directly.
func (es *ExitState) AddPerfData(skipValidate bool, pd ...PerformanceData) error {

	if len(pd) == 0 {
		return ErrNoPerformanceDataProvided
	}

	if !skipValidate {
		for i := range pd {
			if err := pd[i].Validate(); err != nil {
				return err
			}
		}
	}

	es.perfData = append(es.perfData, pd...)

	return nil

}

// AddError appends provided errors to the collection.
func (es *ExitState) AddError(err ...error) {
	es.Errors = append(es.Errors, err...)
}

// SetOutputTarget assigns a target for Nagios plugin output. By default
// output is emitted to os.Stdout.
func (es *ExitState) SetOutputTarget(w io.Writer) {
	// Guard against potential nil argument.
	if w == nil {
		es.outputSink = os.Stdout
	}

	es.outputSink = w
}

// emitOutput writes final plugin output to the previously set output target.
// No further modifications to plugin output are performed.
func (es ExitState) emitOutput(pluginOutput string) {

	// Emit all collected output using user-specified output target. Fall back
	// to standard output if not set.
	if es.outputSink == nil {
		es.outputSink = os.Stdout
	}

	fmt.Fprint(es.outputSink, pluginOutput)
}
